include/deal.II/matrix_free/tensor_product_kernels.h

```diff
@@ -1752,656 +1572,197 @@ namespace internal
             apply_matrix_vector_product<restricted_variant,
                                         quantity,
                                         contract_over_rows,
-                                        add>(
-              shape_data, in, out, n_rows, n_columns, stride, stride);
+                                        add>(shape_data,
+                                             in,
+                                             out,
+                                             n_rows,
+                                             n_columns,
+                                             stride * stride_in,
+                                             stride * stride_out);
 
             if (one_line == false)
               {
-                ++in;
-                ++out;
+                in += stride_in;
+                out += stride_out;
               }
           }
         if (one_line == false)
           {
-            in += stride * (mm - 1);
-            out += stride * (nn - 1);
+            in += stride * (mm - 1) * stride_in;
+            out += stride * (nn - 1) * stride_out;
           }
       }
   }
 
 
 
-  template <EvaluatorVariant variant,
-            int              dim,
-            typename Number,
-            typename Number2>
-  template <int  face_direction,
+  /**
+   * This function applies the tensor product operation to produce face values
+   * from cell values. The algorithm involved here can be interpreted as the
+   * first sweep in sum factorization, reducing the dimensionality of the data
+   * set from dim-dimensional cell values to (dim-1)-dimensional face
+   * values. This step is always done before we evaluate within the face, as
+   * it reduces the length of the loops for the successive steps.
+   *
+   * @tparam n_rows_template The number of entries within the interpolation,
+   *             typically equal to the polynomial degree plus one, if known
+   *             at compile time, otherwise n_rows_runtime is used.
+   * @tparam stride_template The stride between successive entries in the
```

Comment:
more strides: `stride_template` and `stride_runtime`!?

Reply:
Here it is a different role to the stride above. I tried to unify the names with the last commit, and name the stride among the components in the sum factorization operation `stride_operation`.